File: src/utils/anatomia.ts
/**
 * Las cuatro anatomías, y qué acción ofrece cada publicación.
 *
 * El diseño aprobado distingue activo de alto valor, insumo, servicio y
 * logística, y cada una muestra datos distintos. Cuál es no se deduce acá: la
 * declara la publicación en el alta y viaja en la respuesta pública como
 * `operation_kind`. Este módulo sólo traduce esa declaración a lo que la
 * interfaz tiene que dibujar, y es el único lugar donde se decide: si la regla
 * estuviera repartida entre la tarjeta y el detalle, los dos podrían ofrecer
 * acciones distintas sobre la misma publicación.
 */
import { Condition, OperationKind, Product } from '../types';

export type { Condition, OperationKind };

export const ANATOMIAS: OperationKind[] = ['activo', 'insumo', 'servicio', 'logistica'];

/** Las dos que viven en categorías de servicio. */
export const DE_SERVICIO: OperationKind[] = ['servicio', 'logistica'];

export const ETIQUETA_DE_ANATOMIA: Record<OperationKind, string> = {
  activo: 'Activo de alto valor',
  insumo: 'Insumo estandarizado',
  servicio: 'Servicio',
  logistica: 'Logística',
};

/**
 * Lo que dice el backend, o `insumo` si no dice nada.
 *
 * Ausente sólo aparece en datos anteriores a la columna: `insumo` es
 * exactamente cómo se comportaba toda publicación de producto antes —precio,
 * stock y «Agregar»—, así que es la lectura que no cambia nada.
 */
export function normalizarAnatomia(valor?: string | null): OperationKind {
  return (ANATOMIAS as string[]).includes(valor || '')
    ? (valor as OperationKind)
    : 'insumo';
}

export function esDeServicio(anatomia?: OperationKind): boolean {
  return DE_SERVICIO.includes(anatomia || 'insumo');
}

/**
 * ¿Hay un precio para cobrar?
 *
 * No es una interpretación: una publicación sin precio publicado guarda 0, y
 * el catálogo de servicios lo declara además como modalidad `a_convenir`. Lo
 * que no se puede cobrar tampoco puede entrar al carrito —hoy podía, y salía
 * una orden de $0—, y en su lugar dice «A cotizar».
 */
export function tienePrecioPublicado(product: Product): boolean {
  return Number(product.price) > 0;
}

/** Qué puede hacerse con esta publicación, ahora, de verdad. */
export type Accion =
  /** Entra al carrito y sigue por el checkout de siempre. */
  | { tipo: 'comprar'; etiqueta: string }
  /** No hay precio que cobrar: el puente honesto es Contacto. */
  | { tipo: 'cotizar'; etiqueta: string }
  /** Hay precio, pero no queda nada para vender. */
  | { tipo: 'sin-stock'; etiqueta: string };

/**
 * La acción primaria de una publicación.
 *
 * El verbo cambia con la anatomía porque no se compra igual una cosechadora
 * que una bolsa de urea, pero el camino es el mismo que ya existía: carrito y
 * checkout. Lo único que se cierra es comprar algo que no tiene precio.
 */
export function accionDe(product: Product): Accion {
  const anatomia = normalizarAnatomia(product.operationKind);

  if (!tienePrecioPublicado(product)) {
    return { tipo: 'cotizar', etiqueta: 'Solicitar cotización' };
  }

  // Un servicio no tiene unidades que reservar: el backend nunca le descontó
  // stock y por eso no se le pregunta si le queda.
  if (!esDeServicio(anatomia) && !(product.stock > 0)) {
    return { tipo: 'sin-stock', etiqueta: 'Sin stock' };
  }

  switch (anatomia) {
    case 'activo':
      return { tipo: 'comprar', etiqueta: 'Iniciar operación' };
    case 'insumo':
      return { tipo: 'comprar', etiqueta: 'Agregar' };
    default:
      // Servicio y logística con precio publicado: es la contratación que el
      // producto ya soporta, con su carrito y su checkout.
      return { tipo: 'comprar', etiqueta: 'Contratar' };
  }
}

/** La acción secundaria, que siempre nombra el objeto en vez de «ver más». */
export function etiquetaSecundaria(product: Product): string {
  switch (normalizarAnatomia(product.operationKind)) {
    case 'activo':
      return 'Ver condiciones';
    case 'servicio':
      return 'Ver alcance';
    case 'logistica':
      return 'Ver cobertura';
    default:
      return 'Ver detalle';
  }
}

/**
 * Lo que dice el backend sobre la condición, o nada.
 *
 * Nada es una respuesta válida y frecuente: hacienda y campos son activos donde
 * «nuevo o usado» no significa nada, y ninguna publicación anterior a la columna
 * la declaró. Donde falta, la ficha omite la fila. Inventar «nuevo» sería
 * afirmar algo que el vendedor no dijo.
 */
export function normalizarCondicion(valor?: string | null): Condition | undefined {
  return valor === 'nuevo' || valor === 'usado' ? valor : undefined;
}

export const ETIQUETA_DE_CONDICION: Record<Condition, string> = {
  nuevo: 'Nuevo',
  usado: 'Usado',
};

File: backend/app/services/anatomia.py
"""Qué clase de operación es una publicación.

El diseño aprobado pide cuatro anatomías —activo de alto valor, insumo,
servicio y logística— y cada una muestra datos distintos y ofrece una acción
distinta. El producto no tenía con qué elegir entre ellas: `publication_type`
sólo separa producto de servicio, y adivinar el resto por el precio, el título
o el largo de la unidad habría metido una regla invisible en el frontend que
nadie podría corregir después.

Así que la anatomía se declara y se guarda: `products.operation_kind`. Es un
dato de la publicación, no una lectura de la interfaz.

Dos reglas la sostienen:

1. **No contradice al cobro.** Quién paga stock y quién no lo decide desde
   siempre `category.is_service`, y eso no se toca. La anatomía tiene que
   coincidir: `servicio` y `logistica` sólo viven en categorías de servicio, y
   `activo` e `insumo` sólo fuera de ellas. Una publicación que quiera cruzar
   ese límite se rechaza en el alta, no se corrige en silencio.

2. **Nunca queda vacía.** Cada categoría declara su anatomía por omisión, que
   es lo que usa la migración para los registros que ya existían y lo que trae
   preseleccionado el formulario. Es una omisión declarada, no un
   descubrimiento: el dato viejo no distingue una cosechadora de una bolsa de
   urea, y el vendedor puede corregirlo editando la publicación.
"""
from __future__ import annotations

from typing import Optional

# Las cuatro y no hay una quinta.
ACTIVO = "activo"
INSUMO = "insumo"
SERVICIO = "servicio"
LOGISTICA = "logistica"

ANATOMIAS = (ACTIVO, INSUMO, SERVICIO, LOGISTICA)

# Las dos que exigen una categoría de servicio. El complemento exige lo
# contrario: es el mismo límite mirado desde el otro lado.
DE_SERVICIO = (SERVICIO, LOGISTICA)

# Cómo se llaman para un humano. El backend no arma texto de interfaz, pero el
# alta necesita explicar qué está eligiendo el vendedor.
ETIQUETAS = {
    ACTIVO: "Activo de alto valor",
    INSUMO: "Insumo estandarizado",
    SERVICIO: "Servicio",
    LOGISTICA: "Logística",
}

# La anatomía por omisión de cada categoría del catálogo. Se elige por el
# objeto principal de la categoría, no por sus publicaciones de hoy:
# «Repuestos» vende piezas estandarizadas aunque una valga un millón, y
# «Riego y drenaje» vende equipos aunque el envase diga kit.
DEFAULT_POR_CATEGORIA = {
    "maquinaria-agricola": ACTIVO,
    "tierras-parcelas": ACTIVO,
    "bienes-ganado": ACTIVO,
    "ganaderia-forrajes": ACTIVO,
    "riego-drenaje": ACTIVO,
    "agricultura-precision-tecnologia": ACTIVO,
    "insumos-agricolas": INSUMO,
    "repuestos-mantenimiento": INSUMO,
    "asesoramiento": SERVICIO,
    "contratistas": SERVICIO,
    "acopio": SERVICIO,
    "logistica": LOGISTICA,
}


def es_de_servicio(anatomia: Optional[str]) -> bool:
    """¿Esta anatomía vive del lado de los servicios?"""
    return anatomia in DE_SERVICIO


def compatible(anatomia: Optional[str], categoria_es_servicio: bool) -> bool:
    """¿La anatomía cae del mismo lado que la categoría?

    Es la regla que impide que la interfaz prometa una cosa y el cobro haga
    otra: un `insumo` en una categoría de servicio mostraría stock y un botón
    de agregar sobre algo que nunca reserva unidades.
    """
    if anatomia not in ANATOMIAS:
        return False
    return es_de_servicio(anatomia) == bool(categoria_es_servicio)


def por_omision(slug_de_categoria: Optional[str], categoria_es_servicio: bool) -> str:
    """La anatomía que le corresponde a una categoría cuando nadie la declaró.

    Para una categoría que no está en la tabla —una que cree la clienta más
    adelante— se elige la opción que **no cambia el comportamiento actual**:
    hoy toda publicación de producto muestra stock y «Agregar», que es
    exactamente `insumo`; y todo servicio se comporta como `servicio`.
    """
    declarada = DEFAULT_POR_CATEGORIA.get(slug_de_categoria or "")
    if declarada is not None and compatible(declarada, categoria_es_servicio):
        return declarada
    return SERVICIO if categoria_es_servicio else INSUMO


# --- Condición del activo -------------------------------------------------
#
# La anatomía de activo de alto valor pide condición, y el esquema no la tenía:
# «usado» vivía suelto adentro de la descripción, donde no se puede filtrar ni
# comparar. Es obligatoria para publicar un activo nuevo y queda vacía en los
# que ya existían, porque nadie puede saber hoy si aquel tractor era usado sin
# leerle la descripción y adivinar. Donde falta, la ficha omite la fila en vez
# de inventar «nuevo».
NUEVO = "nuevo"
USADO = "usado"

CONDICIONES = (NUEVO, USADO)

ETIQUETA_DE_CONDICION = {
    NUEVO: "Nuevo",
    USADO: "Usado",
}


def condicion_valida(condicion: Optional[str]) -> bool:
    return condicion in CONDICIONES


def usa_condicion(anatomia: Optional[str]) -> bool:
    """¿Esta anatomía muestra condición?

    Sólo el activo de alto valor: un insumo no se compra por su desgaste.

    **No es obligatoria, y eso es una decisión, no un olvido.** `ANATOMIAS.md`
    la pide obligatoria para el activo, pero el catálogo aprobado tiene dos
    categorías de activo donde «nuevo o usado» no significa nada: «Bienes y
    Ganado» —un ternero no es ninguna de las dos— y «Tierras y parcelas».
    Forzarla ahí obligaría al vendedor a elegir una respuesta falsa para poder
    publicar, que es peor que no tener el dato. Se ofrece, se valida si viene,
    y la ficha omite la fila cuando falta.
    """
    return anatomia == ACTIVO
